File: sample/src/main/java/cn/thens/okbinder/sample/OKBinderSample.java
package cn.thens.okbinder.sample;

import android.app.Activity;
import android.app.Service;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Binder;
import android.os.Bundle;
import android.os.IBinder;
import android.os.Parcelable;
import android.util.Log;
import android.widget.Toast;

import cn.thens.okbinder.OkBinder;

public class OKBinderSample {
    private static final String TAG = "@OkBinder";

    interface IRemoteService {
        String test();

        void testError(Boolean aBoolean, Parcelable aParcelable);

        OkBinder<IRemoteService> testCallback(OkBinder<IRemoteService> callback);
    }

    public static class MainActivity extends Activity implements ServiceConnection {
        boolean isServiceConnected = false;

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Toast.makeText(this, "please check the log", Toast.LENGTH_SHORT).show();
            IRemoteService remoteService = OkBinder.client(IRemoteService.class, service);
            Log.d(TAG, "client: test => " + remoteService.test());
            try {
                remoteService.testError(false, ComponentName.unflattenFromString("a.b/.c"));
            } catch (Exception e) {
                Log.e(TAG, "client: testError => \n" + Log.getStackTraceString(e));
            }
            try {
                IRemoteService callback = OkBinder.client(remoteService.testCallback(createRemoteInterface("callback")));
            } catch (Throwable  e) {
                Log.e(TAG, "client: callback error => \n" + Log.getStackTraceString(e));
            }
            Log.d(TAG, "client: callback");
            Log.d(TAG, "end of onServiceConnected");
        }

        @Override
        protected void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            setContentView(R.layout.activity_main);
            findViewById(R.id.vTestLocalService)
                    .setOnClickListener(v -> rebindService(LocalService.class));
            findViewById(R.id.vTestRemoteService)
                    .setOnClickListener(v -> rebindService(RemoteService.class));
        }

        private void rebindService(Class<?> serviceClass) {
            if (isServiceConnected) {
                unbindService(this);
            }
            bindService(new Intent(this, serviceClass), this, Context.BIND_AUTO_CREATE);
            isServiceConnected = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
        }
    }

    public static abstract class BaseService extends Service {
        private Binder okBinder = createRemoteInterface("service");

        @Override
        public IBinder onBind(Intent intent) {
            return okBinder;
        }
    }

    public static class LocalService extends BaseService {
    }

    public static class RemoteService extends BaseService {
    }

    private static OkBinder<IRemoteService> createRemoteInterface(String tag) {
        return OkBinder.server(new IRemoteService() {
            @Override
            public String test() {
                Log.d(TAG, ">> ** " + tag + ": test ** <<");
                return tag + ".test";
            }

            @Override
            public void testError(Boolean aBoolean, Parcelable aParcelable) {
                Log.d(TAG, ">> ** " + tag + ": testError <<");
                throw new NullPointerException();
            }

            @Override
            public OkBinder<IRemoteService> testCallback(OkBinder<IRemoteService> callback) {
                Log.d(TAG, ">> ** " + tag + ": testCallback ** <<");
                IRemoteService remoteInterface = OkBinder.client(callback);
                remoteInterface.test();
                return callback;
            }
        });
    }

}

File: sample/src/main/java/cn/thens/okbinder/sample/OkBinderSample.kt
package cn.thens.okbinder.sample

import android.app.Activity
import android.app.Service
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.os.Parcelable
import android.util.Log
import android.widget.Toast
import cn.thens.okbinder.OkBinder
import kotlinx.android.synthetic.main.activity_main.*

@OkBinder.Interface
interface IRemoteService {
    fun test(): String

    fun testError(aBoolean: Boolean, aParcelable: Parcelable)

    fun testCallback(data: String, callback: ICallback): ICallback
}

@OkBinder.Interface
interface ICallback {
    val data: String
    fun onResult(result: String)
}

private const val TAG = "@OkBinder"

class MainActivity : Activity(), ServiceConnection {
    private var isServiceConnected = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        vTestLocalService.setOnClickListener {
            rebindService(LocalService::class.java)
        }
        vTestRemoteService.setOnClickListener {
            rebindService(RemoteService::class.java)
        }
    }

    private fun rebindService(serviceClass: Class<*>) {
        if (isServiceConnected) {
            unbindService(this)
        }
        bindService(Intent(this, serviceClass), this, Context.BIND_AUTO_CREATE)
        isServiceConnected = true
    }

    override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
        Toast.makeText(this, "please check the log", Toast.LENGTH_SHORT).show()
        val remoteService = OkBinder.proxy(service!!, IRemoteService::class.java)
        remoteService.test().let { Log.d(TAG, "remoteService.test => $it") }
        try {
            remoteService.testError(false, ComponentName.unflattenFromString("a.b/.c")!!)
        } catch (e: Exception) {
            Log.e(TAG, "remoteService.testError => \n" + Log.getStackTraceString(e))
        }
        val callback = remoteService.testCallback("Hello, OkBinder :)", object : ICallback {
            override val data: String = "I_CALLBACK_DATA"

            override fun onResult(result: String) {
                Log.d(TAG, ">> ** ICallback.onResult: result = $result ** <<")
            }
        })
        Log.d(TAG, "callback.data... = ${callback.data}")
        callback.onResult("CALLBACK_RESULT ...")
        Log.d(TAG, "end of onServiceConnected")
    }

    override fun onServiceDisconnected(name: ComponentName?) {
    }
}

abstract class BaseService : Service() {
    private val okBinder = OkBinder(object : IRemoteService {
        override fun test(): String {
            Log.d(TAG, ">> ** IRemoteService.test ** <<")
            return "TEST"
        }

        override fun testError(aBoolean: Boolean, aParcelable: Parcelable) {
            Log.d(TAG, ">> ** IRemoteService.testError: aBoolean = $aBoolean ** <<")
            Log.d(TAG, ">> ** IRemoteService.testError: aParcelable = $aParcelable ** <<")
            throw NullPointerException()
        }

        override fun testCallback(data: String, callback: ICallback): ICallback {
            Log.d(TAG, ">> ** IRemoteService.testCallback: data = $data ** <<")
            Log.d(TAG, ">> ** IRemoteService.testCallback: callback.data = ${callback.data} ** <<")
            callback.onResult("CALLBACK_RESULT")
            return callback
        }
    })

    override fun onBind(intent: Intent?): IBinder? {
        return okBinder
    }
}

class LocalService : BaseService()

class RemoteService : BaseService()